Fixes #15411 - Move smart_proxy_ids_for to Host

xprazak2 opened the pull request.

Commit:
Fixes #15411 - Move smart_proxy_ids_for to Host

ShimShtein commented:
Thanks for helping me out with this! I really appreciate it. Unfortunately the issue in redmine was unclear, so I'll try to explain it here:

First, my end goal is to move puppet to a plugin. As a preparation steps, I am trying to organize the core in such manner that the moving to a plugin will be easier.
Now for our specific issue:
In core we had a method in `SmartProxy` class that was influenced by logic from host - it knew which fields in host object contained a reference to a `SmartProxy` object. I wanted to move the method to a host object (where it belongs).
Now about making the method more extensible: I had in mind that even the host object itself doesn't have all the information where to search for `SmartProxy` instances, for example the lines that add a reference to `puppet_proxy_id` and `puppet_ca_proxy_id` should be added from outside the core object.

Let me make a proposal on how I would implement this:
First we have to prepare the host object: we will add a `belongs_to` association for each `SmartProxy` reference:

``` ruby
class Host::Managed
  belongs_to :puppet_proxy, :class => 'SmartProxy'
end
```

Next we will add DSL that will signal that this relation is a relation to `SmartProxy`:

``` ruby
class Host::Managed
  smart_proxy_reference :puppet_proxy
end
```

And then we have all the information needed for implementing `#smart_proxy_ids` method:

``` ruby
def self.smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
  id_columns = smart_proxy_references.map { |r| "#{r}_id" }
  hosts_scope.eager_load(smart_proxy_references).pluck(id_columns).flatten.compact.uniq
end
```

This new method will be even more efficient, since it will run only one SQL query (with a million joins though).

If you think it's too much hassle than you planned to spend on this, just ping me, I'll take over.

xprazak2 commented:
I made changes as suggested. We track the references to proxies through `smart_proxy_reference`. I have to do the joins in sequence to avoid name ambiguity.

ShimShtein commented on test/unit/host_test.rb:
```diff
@@ -3239,6 +3239,28 @@ def to_managed!
+    puppet_id = host_5.puppet_proxy_id
+    puppet_ca_id = host_6.puppet_ca_proxy_id
+
+    res = Host.smart_proxy_ids(Host.where(:id => [host_1, host_2, host_3, host_4].map(&:id)))
```
Please add host_5 and host_6 to the list, I think it's the reason the test doesn't pass.

ShimShtein commented:
Generally looks fantastic, can you please address the rubocop and testing issues?

Thanks!

ShimShtein commented:
I wonder if this code will be more efficient for `smart_proxy_ids`:

``` ruby
def self.smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
  hosts_scope.eager_load(smart_proxy_references.map { |r| r[:ref] }).map( |host| smart_proxy_references.map { |r| host.send(r[:ref]).try(:id) } ).flatten.uniq.compact
end
```

Actually then we can get rid of `:column` completely from `smart_proxy_references`.

Update: It will look even more readable:

``` ruby
def self.smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
  hosts_scope.eager_load(smart_proxy_references).map( |host| smart_proxy_references.map { |r| host.send(r).try(:id) } ).flatten.uniq.compact
end
```

tbrisker commented on app/models/concerns/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,12 @@
+module SmartProxyReference
```
If I understand your intention here correctly, there is something very similar in the Hostext::Search concern that is used for searching hosts by proxy - https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/search.rb#L233
If so, these should be refactored so we have one source of truth about what are all the hosts' relations with proxies.

ShimShtein replied:
Thanks for pointing this out. Indeed it's similar, just the query goes the other way around - in our case we start with hosts and collect smart proxy references, in hostext you start with a list of smart proxies and collect host ids.

You are right that the current code in hostext needs refactoring, since it counts on the fact that `eager_load` won't alias table names. It is true only to the point when you reference the same table twice, if you do have such reference - `eager_load` will create aliases to those tables that are not very predictable.

We can use the same concept on the smart proxy - create a set of `has_one`/`belongs_to` statements in `SmartProxy` class and mark them in the same way as we do here. If we want to force a single source of truth, we can force in the DSL to mark both ways of the connection, something along the lines of:

``` ruby
smart_proxy_reference :host => :puppet_proxy, :proxy => :puppet_host
```

ShimShtein commented on app/models/host/managed.rb:
```diff
@@ -231,6 +239,12 @@ def self.model_name
 
+  def self.smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+    hosts_scope.eager_load(smart_proxy_references).map do |host|
+      smart_proxy_references.map { |ref| host.send(ref).try(:id) }.flatten.uniq.compact
```
The `.flatten.uniq.compact` should be after the `end` in the next line I think.
If I read the code correctly, line 244 produces an array of proxy ids for a single host `[2,3,nil, nil,3]` and the `map` on line 243 aggregates it to a single two dimensional array: `[[2,3,nil, nil,3], [1,2,nil, 1,1], [nil,nil,nil, nil,5]]`. This array should be flattened, uniqued and compacted.

tbrisker replied:
`.flat_map` would be much more efficient then `.map{...}.flatten` if there is only one level of nesting.

xprazak2 commented:
I made changes in `Hostext::Search` so we have one source of truth for relationships with proxies. I replaced `map` with `flat_map` and moved the misplaced `.uniq.compact`. I left the query as is for now.

ShimShtein commented:
After a long discussion with @tbrisker, we came to an agreement as for the best way of implementing the query. I'll start with listing the constraints that we identified:
1. We have to use a single query (for performance sake)
2. We don't want to construct the `SmartProxy` objects that we want to enumerate.
3. We don't know how many times each table will be referenced.

As a conclusion, this means that we have to use `#pluck` for getting the ids without the record. On the brighter side, we can get rid of the new relations we have added (like `:domain_dns`) since we can get the proxy id by selecting `dns_id` column in parent (`:domain` for example) relation. Now we are getting to the less pleasant part of this: `#pluck` does not work with relation names - it can work only with symbols (if the column is in the main table) or string of type `'table_name.column_name'`. This leads us to changing the registry again (sorry about going back and forth with that...) to store both relation and a list of columns containing proxy ids there.
Example:

``` ruby
{
  :subnet => [:dns_id, :dhcp_id, :tftp_proxy_id],
  :subnet6 => [:dns_id, :dhcp_id, :tftp_proxy_id]
}
```

Now we have all the information to construct a query that we want:

``` ruby
base_scope.eager_load(registry.keys).pluck(create_column_list(registry)).flatten.uniq.compact
```

All that is left, is to define the `create_column_list` method, but here comes the complex thing: `#eager_load` will give an alias to a table, if the the table appears more than once in the final join. In the example I shown earlier, we have both `subnet` and `subnet6` that both resolve to `subnets` table. Here we have to mimic the algorithm that is used by the `#eager_load`: the first appearance of a table will not get an alias (it will be called `subnets` in our case); each of the following appearances will get an alias constructed from plural form of the _relation_ concated with plural name of relation's active record table name (in our example it would be `subnet6s_hosts`). Lucky for us, all this information is available in `Host::Managed.reflections` method. The base algorithm for `create_column_list` should go something like this:
1. group all relations in the registry by `Host::Managed.reflections[relation_key].table_name`
2. Iterate through the groups: for first relation generate a list of strings with the pattern `"#{relation_name}.#{column_name}"` for each column in the registry. For the next relations in the group, the string should look like this `"#{reflection.plural_name}_hosts.#{column_name}"` for each column in the registry.

Now about `Hosttext::Search` code: the code there should use the `create_column_list` too, but in a bit different context. The code there doesn't need to pluck the values of those columns, but create a join with filtered `smart_proxies` table, where the `ON` statement should be our list of columns:

``` ruby
host_ids = Host::Managed.reorder('')
                                .authorized(:view_hosts, Host)
                                .eager_load(Host.smart_proxy_references)
                                .joins("LEFT JOIN smart_proxies ON smart_proxies.id IN (#{create_column_list(registry)})")
                                .where(proxy_cond)
                                .uniq
                                .pluck('hosts.id')
```

Thanks for staying with me for this long :)

xprazak2 commented:
No problem :-), now it should be as described above, with plucking the ids after we renamed the duplicate tables if there were any

ShimShtein commented on app/models/concerns/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,51 @@
+
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...]}")) unless correct_reference_format?(hash)
+      @smart_proxy_references ||= {}
```
IMHO The code would be a bit more readable if we use `OpenStruct` instead of plain hash as the internal storage.
I won't fight over it though...

xprazak2 replied:
If you don't mind, I would like to keep just a simple hash because `OpenStruct` is not enumerable I can use `select` or `reject` to find columns on `self` easily.

ShimShtein commented on app/models/concerns/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,51 @@
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...]}")) unless correct_reference_format?(hash)
+      @smart_proxy_references ||= {}
+      @smart_proxy_references.merge!(hash)
```
This code prevents adding fields from the same relation in different statements:
Consider:

``` ruby
smart_proxy_reference :subnets => [:dhcp_id, :dns_id]
#then later in code (maybe in a different plugin)
smart_proxy_reference :subnets => [:tftp_id]
```

`:dhcp_id` and `:dns_id` would be removed, since `merge` doesn't deal with array values.

ShimShtein commented on app/models/concerns/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,51 @@
+          relation.map do |rel_name, columns|
+            columns.map do |col_name|
+              rename = i > 0 ? "_hosts" : ""
+              "#{rel_name.pluralize}#{rename}.#{col_name}"
```
In case of the first relation, table name would be taken, not the relation one:

``` ruby
alias_name = i > 0 ? "#{rel_name.pluralize}_hosts" : table_name
"#{alias_name}.#{col_name}"
```

ShimShtein commented:
@xprazak2 Looks great! Two small comments plus a rebase and I think it's ready.

xprazak2 commented:
[test]

ShimShtein commented:
@tbrisker Now it looks good to me, can you PLZ review it again?

ShimShtein commented:
@xprazak2, can you rebase this PLZ?

dLobatog commented:
What's the status here?

xprazak2 commented:
Rebased to fix conflicts.

xprazak2 commented:
[test foreman]

ohadlevy commented:
@xprazak2 what is the status of this PR? thanks

xprazak2 commented:
I rebased, the query is constructed as per @ShimShtein's comments above, ready for another round of reviews.

ShimShtein commented:
Small change is required to adhere to our registry policy:

`app/models/concerns/smart_proxy_reference.rb` should be renamed to `SmartProxyHostExtensions`

Once we have `SmartProxyReference` free, we can add a new file: `app/services/smart_proxy_reference.rb` and put the registry part there. The overall result should be similar to the one we have [here](https://github.com/theforeman/foreman/blob/develop/app/services/host_info.rb):

- [ ] `self.smart_proxy_ids` should be moved from `Host::Managed` to `SmartProxyHostExtensions` (to encapsulate smart proxy business to the concern.
- [ ] `smart_proxy_references` attribute should be moved to our new registry file
- [ ] `def smart_proxy_reference(hash)` method should be moved to our new registry file
- [ ] make `def smart_proxy_reference(hash)` in our registry to do validation check and return a newly created hash (`:table => [ :field1, :field2]`)
- [ ] add a new attribute `to `plugin.rb` [class](https://github.com/theforeman/foreman/blob/develop/app/registries/foreman/plugin.rb#L99)
- [ ] add a `smart_proxy_reference` method to `plugin.rb` that will call our registry's `smart_proxy_reference` and accumulate its result into the newly created variable.
- [ ] add two methods to the registry class: `local_entries` and `plugin_entries` similar to what we have in [host_info](https://github.com/theforeman/foreman/blob/develop/app/services/host_info.rb#L12-L18).
- [ ] add `add_smart_proxy_reference` method to registry class that will call `smart_proxy_reference` and store the result into `local_entries`.
- [ ] add `smart_proxy_references` (plural) method to the registry that will gather all information from `local_entries` + `plugin_entries` and generate a single hash that can (and should) be used by `SmartProxyExtensions#proxy_join_references`.

Hopefully it will be the last change. All in all it looks great!

Thanks again for getting us this far!

xprazak2 commented:
I made changes according to the steps outlined above, I hope I did not miss anything.

ShimShtein commented on app/models/host/managed.rb:
```diff
@@ -67,6 +67,12 @@ def build_hooks
 
   include HostCommon
 
+  SmartProxyReference.smart_proxy_reference :subnet => [:dns_id, :dhcp_id, :tftp_id]
```
We can add:
``` ruby
def smart_proxy_reference(hash)
  SmartProxyReference.add_smart_proxy_reference(hash)
end
```
to `smart_proxy_host_extensions.rb`. It will make declarations even prettier.

ShimShtein commented on app/registries/foreman/plugin.rb:
```diff
@@ -412,6 +413,10 @@ def action_scopes_hash_for(controller_class)
     end
 
+    def smart_proxy_reference(hash)
+      @smart_proxy_references.smart_proxy_reference(hash)
```
Probably typo, should be:
``` ruby
@smart_proxy_references << SmartProxyReference.smart_proxy_reference(hash)
```

also, could you please add a unit test for the plugin functionality, like [here](https://github.com/theforeman/foreman/blob/develop/test/unit/plugin_test.rb#L402)?

ShimShtein commented on app/services/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    def entries_from_plugins
+      Foreman::Plugin.all.flat_map(&:smart_proxy_references).compact.reduce(:merge)
```
This can cause problems, since it overwrites the column list:
``` ruby
[{:self => [:id] },{ :self => [:name]}].reduce(:merge)
# => {:self=>[:name]}
```

ShimShtein commented on test/unit/concerns/smart_proxy_reference_test.rb:
```diff
@@ -0,0 +1,17 @@
+
+class SmartProxyReferenceTest < ActiveSupport::TestCase
+  class ProxyReferrer
+    include SmartProxyReference
```
I think this is not the intended usage. Should be changed to `SmartProxyHostExtensions`.

ShimShtein reviewed (changes requested):
Almost there!
A couple of comments inline though.

tbrisker commented on app/services/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,51 @@
+    end
+
+    def entries_from_plugins
+      Foreman::Plugin.all.flat_map(&:smart_proxy_references).compact.reduce({}) do |memo, hash|
```
disregard, we need to disable this cop...

xprazak2 commented:
I made the declarations prettier, fixed the typos and added a couple of tests. I am not entirely sure how the new/lazy populated registry type is meant to isolate it's state for tests, so I went for the simplest solution I could think of.

ShimShtein commented on test/unit/plugin_test.rb:
```diff
@@ -479,6 +479,16 @@ def test_add_dashboard_widget
+      smart_proxy_reference :hosts => [:test]
+    end
+    assert_equal [:test], SmartProxyReference.entries_from_plugins[:hosts]
+    SmartProxyReferences.references = refs
```
can you put this into `ensure` block? If the test fails, it causes other tests to behave unxpectedly.

ShimShtein commented on test/unit/smart_proxy_reference_test.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  test "should add correct entries from plugins" do
+    Foreman::Plugin.register :test_first_entry_from_plugin do
```
Don't we need to clean up plugin registry too?

ShimShtein reviewed (changes requested):
2 more comments.
Could you also look at the failing tests?

xprazak2 commented:
I put the cleaning into `ensure` block where needed and added a method to remove a plugin from registered plugins to keep the `@smart_proxy_references` for a plugin read-only.

ShimShtein commented on test/unit/smart_proxy_reference_test.rb:
```diff
@@ -0,0 +1,36 @@
+    begin
+      assert_equal [:my_test, :my_test_again], SmartProxyReference.smart_proxy_references[:hosts].sort
+    ensure
+      Foreman::Plugin.unregister :test_first_entry_from_plugin
```
I would prefer not introducing a public method to unregister a plugin - we can't promise that it will remove all the leftovers from that plugin.

We can isolate plugins list very similar to what we have done with `SmartProxyReferences.references`:
``` ruby
old_plugins = Foreman::Plugin.registered_plugins
# Do test
ensure
  Foreman::Plugin.instance_variable_set(old_plugins)
end
```
Another option would be using mocha for that:

``` ruby
setup do
  temp_plugin_list = Foreman::Plugin.registered_plugins.clone
  Foreman::Plugin.stubs(:registered_plugins).returns(temp_plugin_list)
end
```

ares commented:
seems that tests are failing

mmoll commented:
[test foreman]

xprazak2 commented:
[test foreman]
[test katello]

xprazak2 commented:
@ShimShtein, I finally fixed the tests, could you give this another look please?

ShimShtein commented:
:+1:  from me. Looks great, and works as expected!

Thanks!

@tbrisker, could you please give this another review round?

mmoll commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    def smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+      hosts_scope.eager_load(proxy_join_tables).pluck(proxy_column_list).flatten.uniq.compact
```
can `distinct` be used here instead of `uniq`? see GH-4838

xprazak2 commented:
I replaced `uniq` with `distinct`

mmoll commented:
[test katello]

mmoll commented:
OK, seems that `uniq` was already too late and thus already applied on a Ruby object and not on the query...

xprazak2 commented:
I reverted back to `uniq`

mmoll commented:
[test katello]

iNecas commented:
This needs a reabse to get Katello tests green

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+      proxy_join_references.keys
+    end
+
+    def create_groups
```
Pretty late in the review party, but since the module is getting included in to `Host` class, this name can be pretty misleading in the context of the host. Could you think of a better name?

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+    def create_groups
+      refs = proxy_join_references.inject([]) { |memo, (k, v)| memo << { k => v } }
+      refs.select { |ref| host_reflection(ref.keys.first) }
+        .group_by { |h| host_reflection(h.keys.first).table_name }.with_indifferent_access
```
The `ref` argument in the line 23 and `h` in 24 represent the same type of items, right? Using different names makes it harder to read the code

iNecas commented on app/services/smart_proxy_reference.rb:
```diff
@@ -0,0 +1,51 @@
+      join_references result, entries_from_plugins
+    end
+
+    def local_entries
```
why calling `local_entries` the method that returns `referencies` Shouldn't we either rename one or the other to match each other?

xprazak2 commented:
I rebased and improved style as suggested.

mmoll commented:
test failures unrelated.

ares commented:
@iNecas any final comments?

iNecas commented:
Yes, I want to chat about some part of the codes with @xprazak2 to help me understand how that actually works and if there are some ways to improve the readability.

dLobatog reviewed (changes requested):
Set to WoC after @iNecas comment, @xprazak2 let us know when it's ready

ares commented:
@inecas did you chat with @xprazak2? It seems to got stuck.

xprazak2 commented:
[test foreman]

mmoll commented:
this needs a rebase for the tests to complete

xprazak2 commented:
rebased

dLobatog reviewed (dismissed):
@xprazak2 Check out tests

ares commented:
@xprazak2 this needs another rebase :-( before you do it, I suggest you grab @iNecas and do a peer review with him, so we don't waste cycles only on rebasing

mmoll commented:
@xprazak2 please rebase :)
@iNecas could you have another look here?

xprazak2 commented:
Rebased

mmoll commented on test/models/host_test.rb:
```diff
@@ -3591,6 +3591,28 @@ def to_managed!
+    host_3 = FactoryGirl.create(:host, :with_dhcp_orchestration)
+    host_4 = FactoryGirl.create(:host, :with_realm)
+    host_5 = FactoryGirl.create(:host, :with_puppet)
+    host_6 = FactoryGirl.create(:host, :with_puppet_ca)
```
FactoryBot

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+      proxy_join_references.keys
+    end
+
+    def create_proxy_reference_groups
```
I guess this method should be private

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def smart_proxy_reference(hash)
```
This and other methods would deserve some documentation on what options they expect and how they achieve the stuff, since they use quite a lot of reflections and other dark magic, it's hard to read and understand the code (which might be a reason why I might be unconsciously avoiding reviewing the PR :)

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    def proxy_column_list
+      columns = create_proxy_reference_groups.map do |table_name, relations|
```
I guess this is the most complex method in the code, that definitely deserves either extracting some parts for readability reasons, and/or some additional explanational comments.

Perhaps one of the reasons for the readability issues is the fact, that we're using a hash as the basic data structure. I wonder if replacing that with actual `SmartProxyReference` objects, and perhaps extracting some code there, would help?

xprazak2 commented:
Updated, I added the `SmartProxyReference` object that represents the actual reference and replaces hash for keeping track of references. I moved some code in there in an attempt to make it more readable.

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,51 @@
+      proxy_self_references.map(&:columns_to_s)
+    end
+
+    def reference_groups_by_table
```
I guess we should include `proxy_` prefix for this method as well, to make it clear it's not about generic refernce

iNecas commented on app/models/concerns/smart_proxy_host_extensions.rb:
```diff
@@ -0,0 +1,51 @@
+    def reference_groups_by_table
+      join_references_with_reflection.group_by(&:table_name).with_indifferent_access
+    end
+
```
Same here: let's use `proxy_` prefix in the method name

iNecas commented:
Test failures related. I had one comment regarding the method perfix, but after that and tests passing, good to go from my side

xprazak2 commented on app/services/proxy_reference_registry.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    def join_reference_arrays(references, other_references)
+      to_add = other_references.reduce([]) do |memo, other_ref|
```
Really?

xprazak2 commented:
Updated, I added prefix to methods and `unregister` method for plugins which is convenient in tests

mmoll commented:
[test upgrade]

iNecas reviewed (dismissed):
Tested and works as expected. Waiting for tests to finish.

mmoll commented:
@xprazak2 please check the test failures :point_up:

mmoll commented:
merged, díky @xprazak2!

lzap commented:
We see random failures in core around proxy tests last week, I wonder if this patch affected it in any way: https://community.theforeman.org/t/random-failures-in-core/8852
